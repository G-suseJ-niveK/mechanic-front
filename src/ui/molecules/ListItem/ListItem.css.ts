import { makeStyles, Theme } from '@material-ui/core/styles';

const useStyles = makeStyles((theme: Theme) => ({
  nested: {
    paddingLeft: theme.spacing(4)
  },
  iconList: {
    '& .MuiListItemIcon-root': {
      minWidth: '35px'
    }
  },
  ItemOfListActive: {
    background: '#96c26299 !important',
    boxShadow: '4px 4px 10px rgb(0 0 0 / 15%)',
    borderRadius: '5px',
    '&:hover': {
      background: '#96c26299 !important'
    }
  },
  ItemOfListHover: {
    background: 'none',
    '&:hover': {
      background: '#f5e27f61'
    }
  }
}));

export default useStyles;
