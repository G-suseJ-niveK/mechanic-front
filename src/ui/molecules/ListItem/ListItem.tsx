import React, { useState, useCallback } from 'react';
import useStyles from './ListItem.css';
import { ListItem, ListItemIcon, Icon, ListItemText, Collapse, List } from '@material-ui/core';
import { ExpandLess, ExpandMore } from '@material-ui/icons';
import { NavLink } from 'react-router-dom';
import { useMediaQuery } from '@material-ui/core';
import { useTheme } from '@material-ui/core/styles';

type ListItemProps = {
  item: any;
  onClick?(event: React.MouseEvent<HTMLButtonElement, MouseEvent>): void;
};

const ComponentListItem: React.FC<ListItemProps> = (props: ListItemProps) => {
  const themes = useTheme();
  const [open, setOpen] = useState<boolean>(false);
  const classes = useStyles();
  const { item, onClick } = props;
  const isActiveDesktop = useMediaQuery(themes.breakpoints.up('md'));

  const handleShowChildren = () => {
    setOpen((prevValue: boolean) => !prevValue);
  };

  const handleOnClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
      if (!isActiveDesktop) {
        onClick && onClick(event);
      }
    },
    [onClick, isActiveDesktop]
  );

  // const getActiveLink = useCallback((route: string, currentRoute: string): boolean => {
  //   if (route.toLocaleLowerCase() === currentRoute.toLocaleLowerCase()) {
  //     return true;
  //   }
  //   return false;
  // }, []);

  return (
    <div style={{ fontSize: '2em' }}>
      {item.children && item.children.length > 0 ? (
        <>
          <ListItem
            button
            onClick={handleShowChildren}
            component={React.forwardRef((props: any, ref: any) => (
              <NavLink
                // eslint-disable-next-line react/prop-types
                to={item.path}
                {...props}
                ref={ref}
                exact
                activeClassName={classes.ItemOfListActive}
              />
            ))}
            className={classes.ItemOfListHover}>
            <ListItemIcon style={{ minWidth: '35px' }}>
              <Icon> {item.icon} </Icon>
            </ListItemIcon>
            <ListItemText primary={item.text} style={{ color: '#FFF' }} />
            {open ? <ExpandLess style={{ color: '#FFFFFF' }} /> : <ExpandMore style={{ color: '#FFFFFF' }} />}
          </ListItem>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <List component="div" disablePadding>
              {item.children.map((child: any, index: number) => {
                return (
                  // <ListItem button className={classes.nested} >
                  //   <ListItemIcon>
                  //     <Icon style={{ color: themes.sidebar.iconColor }}>{child.icon}</Icon>
                  //   </ListItemIcon>
                  //   <ListItemText primary={child.text} />
                  // </ListItem>
                  <ListItem
                    button
                    onClick={handleOnClick}
                    key={`list_item_child${item.text}_${index}`}
                    component={React.forwardRef((props: any, ref: any) => (
                      <NavLink
                        // eslint-disable-next-line react/prop-types
                        to={child.path}
                        {...props}
                        ref={ref}
                        exact
                        activeClassName={classes.ItemOfListActive}
                      />
                    ))}
                    className={classes.nested}>
                    <ListItemIcon>
                      <Icon>{child.icon}</Icon>
                    </ListItemIcon>
                    <ListItemText primary={child.text} />
                  </ListItem>
                );
              })}
            </List>
          </Collapse>
        </>
      ) : (
        <ListItem
          button
          onClick={handleOnClick}
          component={React.forwardRef((props: any, ref: any) => (
            <NavLink
              // eslint-disable-next-line react/prop-types
              to={item.path}
              {...props}
              ref={ref}
              exact
              activeClassName={classes.ItemOfListActive}
            />
          ))}
          className={classes.ItemOfListHover}>
          <ListItemIcon>
            <Icon>{item.icon} </Icon>
          </ListItemIcon>
          <ListItemText primary={item.text} />
        </ListItem>
      )}
    </div>
  );
};

export default React.memo(ComponentListItem);
