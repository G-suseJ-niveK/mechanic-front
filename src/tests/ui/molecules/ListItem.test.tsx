import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import ListItem from '~molecules/ListItem/ListItem';
import { BrowserRouter } from 'react-router-dom';

describe('Render ListItem', () => {
  const item = {
    text: 'mock-text',
    path: 'mock-path',
    icon: 'mock-icon'
  };

  const items = {
    text: 'mock-text',
    path: 'mock-path',
    icon: 'mock-icon',
    children: [
      { text: 'mock-text-1', path: 'mock-path-1', icon: 'mock-icon-1' },
      { text: 'mock-text-2', path: 'mock-path-2', icon: 'mock-icon-2' }
    ]
  };

  test('should render', () => {
    const { container } = render(
      <BrowserRouter>
        <ListItem item={item} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test('should render children', () => {
    render(
      <BrowserRouter>
        <ListItem item={items} />
      </BrowserRouter>
    );
    fireEvent.click(screen.getByRole('button'));
    expect(screen.getByRole('button', { name: 'mock-text-1' })).toBeInstanceOf(Node);
    expect(screen.getByRole('button', { name: 'mock-text-2' })).toBeInstanceOf(Node);
  });
});
