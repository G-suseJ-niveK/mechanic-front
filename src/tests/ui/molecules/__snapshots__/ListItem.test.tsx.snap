// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ListItem should render 1`] = `
<div>
  <div
    style="font-size: 2em; background: rgb(33, 64, 54);"
  >
    <a
      aria-disabled="false"
      class="MuiButtonBase-root MuiListItem-root makeStyles-ItemOfListHover-4 MuiListItem-gutters MuiListItem-button"
      href="/mock-path"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiListItemIcon-root"
      >
        <span
          aria-hidden="true"
          class="material-icons MuiIcon-root"
          style="color: rgb(138, 184, 51);"
        >
          mock-icon
           
        </span>
      </div>
      <div
        class="MuiListItemText-root"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          mock-text
        </span>
      </div>
      <span
        class="MuiTouchRipple-root"
      />
    </a>
  </div>
</div>
`;
